### PR TITLE
Add `emcc` toolchain for WASM builds using Emscripten

### PR DESCRIPTION
**Pull request by daviwil**
First of all, thanks so much for creating this project, it fits my requirements perfectly!  I wanted a better way to build widely cross-platform apps without relying on CMake, etc, and I happened to find XMake and xmake.sh a few weeks ago.  Great work!

This PR updates `path_toolname` to detect the names of the Emscripten tools so that `configure` can be used with the `emconfigure` command.  After these changes (and installing the Emscripten SDK), one can build their project for the web by using the following commands:

```
emconfigure ./configure
make
```

I'm not sure if a full toolchain description should be added for Emscripten instead, but these minimal changes seemed to be enough to get everything working perfectly on my end.

### CODE REVIEW

**Comment by waruqi** [comment]
We should add a toolchain named (emcc). like this https://github.com/xmake-io/xmake.sh/blob/6b1fa8e26371ce282d5564607715e563e6b9f27b/configure#L2840

and add a new platform and arch, `wasm` and `wasm32`

then run 

```
./configure --plat=wasm --arch=wasm32
make
```

and we can improve to detect platform/arch in emcc envs.

so we just need 

```
./configure
make
```

**Comment by daviwil** [comment]
Added a new toolchain called `wasm32_unknown_wasm`, I'm not sure if this is the correct triplet name so please let me know if I should change it.

File: configure [hunk 1]
```diff
@@ -2878,6 +2878,18 @@ toolchain "aarch64_linux_gnu"
     set_toolset "ar" "aarch64-linux-gnu-ar" "ar"
 toolchain_end
 
+# emcc toolchain (wasm32)
+toolchain "wasm32_unknown_wasm"
```
Comment (waruqi) [inline configure]: please rename to `emcc` toolchain. Consistent with xmake's toolchain naming. https://github.com/xmake-io/xmake/tree/master/xmake/toolchains/emcc

File: configure [hunk 2]
```diff
@@ -409,9 +409,15 @@ path_toolname() {
         clang) toolname="clang";;
         clang-*) toolname="clang";;
         */clang-*) toolname="clang";;
+        */emcc) toolname="clang";;
```
Comment (waruqi) [inline configure]: and it should be `toolname = emcc`

File: configure [hunk 3]
```diff
@@ -409,9 +409,15 @@ path_toolname() {
...
+        emcc) toolname="clang";;
+        */em++) toolname="clangxx";;
+        em++) toolname="clangxx";;
```
Comment (waruqi) [inline configure]: `toolname="emxx"`

File: configure [hunk 4]
```diff
@@ -409,9 +409,15 @@ path_toolname() {
...
         *-ar) toolname="ar";;
         */ar) toolname="ar";;
         ar) toolname="ar";;
+        */emar) toolname="ar";;
+        emar) toolname="ar";;
```
Comment (waruqi) [inline configure]: emar

**Comment by waruqi** [comment]
and you can check envs to detect wasm platform and arch automatically.

https://github.com/xmake-io/xmake.sh/blob/71223883bd63f5993d46b4de3bf127a295591fb6/configure#L637

**Comment by daviwil** [comment]
> and you can check envs to detect wasm platform and arch automatically.

Do you have a suggestion for which environment variable to check?  It looks like `EMMAKEN_JUST_CONFIGURE=1` is set when `emconfigure` is run, that one might be useful.  I suppose we could also just check for `EMSCRIPTEN` too.  Which do you prefer?

File: configure (hunk 2 above)
Comment (daviwil) [reply configure]: Added all emcc toolname references.

File: configure (hunk 1 above)
Comment (daviwil) [reply configure]: Done.

**Comment by daviwil** [comment]
Looks like `EMSDK` is the only reliable variable present when the Emscripten SDK shell has been loaded up, so I'm checking for that.  Seems to work as expected when I run this:

```sh
docker run -v `pwd`:`pwd` -w `pwd` -u $(id -u):$(id -g) emscripten/emsdk /bin/sh -c "./configure"
```

Output:

```
checking for platform ... wasm
checking for architecture ... x86_64
checking for make ... ok
checking for the c (cc) ... emcc
checking for the c++ (cxx) ... emcc
checking for the assembler (as) ... emcc
checking for the objc (mm) ... emcc
checking for the objc++ (mxx) ... emcc
checking for the linker (ld) ... em++
checking for the static (ar) ... emar
checking for the shared (sh) ... em++
checking for toolchain ... emcc
...
```

**Comment by waruqi** [comment]
ok, thanks!

**Comment by daviwil** [comment]
Thank you, too!

**Pull request merged**